src/core/featurehistory.cpp

```diff
@@ -0,0 +1,448 @@
+    disconnect( vl, &QgsVectorLayer::committedFeaturesAdded, this, &FeatureHistory::onCommittedFeaturesAdded );
+    disconnect( vl, &QgsVectorLayer::committedFeaturesRemoved, this, &FeatureHistory::onCommittedFeaturesRemoved );
+
+    connect( vl, &QgsVectorLayer::beforeCommitChanges, this, &FeatureHistory::onBeforeCommitChanges );
+    connect( vl, &QgsVectorLayer::afterCommitChanges, this, &FeatureHistory::onAfterCommitChanges );
+    connect( vl, &QgsVectorLayer::committedFeaturesAdded, this, &FeatureHistory::onCommittedFeaturesAdded );
+    connect( vl, &QgsVectorLayer::committedFeaturesRemoved, this, &FeatureHistory::onCommittedFeaturesRemoved );
+
+    mObservedLayerIds.insert( vl->id() );
+  }
+}
+
+
+void FeatureHistory::onBeforeCommitChanges()
+{
+  if ( mIsApplyingModifications )
+  {
+    return;
+  }
+
+  QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( sender() );
+
+  if ( !vl )
+  {
+    return;
+  }
+
+  QgsVectorLayerEditBuffer *eb = vl->editBuffer();
+
+  if ( !eb )
+  {
+    return;
+  }
+
+  const QgsFeatureIds deletedFids = eb->deletedFeatureIds();
+  const QgsFeatureIds changedGeometriesFids = qgis::listToSet( eb->changedGeometries().keys() );
+  const QgsFeatureIds changedAttributesFids = qgis::listToSet( eb->changedAttributeValues().keys() );
+  // NOTE QgsFeatureIds underlying implementation is QSet, so no need to check if the QgsFeatureId already exists
+  QgsFeatureIds changedFids;
+
+  for ( const QgsFeatureId fid : deletedFids )
+    changedFids.insert( fid );
+
+  for ( const QgsFeatureId fid : changedGeometriesFids )
+    changedFids.insert( fid );
+
+  for ( const QgsFeatureId fid : changedAttributesFids )
+    changedFids.insert( fid );
+
+  // NOTE we read the features from the dataProvider directly as we want to access the old values.
+  // If we use the layer, we get the values from the edit buffer.
+  QgsFeatureIterator featuresIt = vl->dataProvider()->getFeatures( QgsFeatureRequest( changedFids ) );
+  QMap<QgsFeatureId, QgsFeature> modifiedFeatures;
+  QgsFeature f;
+
+  // ? is it possible to use the iterator in a less ugly way? something like normal `for ( QgsFeature &f : it ) {}`
+  while ( featuresIt.nextFeature( f ) )
+    modifiedFeatures.insert( f.id(), f );
+
+  qInfo() << "FeatureHistory::onBeforeCommitChanges: vl->id()=" << vl->id() << "sourcePkAttrPair=" << changedFids;
+
+  // NOTE no need to keep track of added features, as they are always present in the layer after commit
+  mTempModifiedFeaturesByLayerId.insert( vl->id(), modifiedFeatures );
+}
+
+
+void FeatureHistory::onCommittedFeaturesAdded( const QString &localLayerId, const QgsFeatureList &addedFeatures )
+{
+  if ( mIsApplyingModifications )
+  {
+    return;
+  }
+
+  const QgsVectorLayer *vl = qobject_cast<QgsVectorLayer *>( sender() );
+
+  if ( !vl )
+  {
+    return;
+  }
+
+  qInfo() << "FeatureHistory::onCommittedFeaturesAdded: adding create committed features";
```

Comment:
I'm not sure we want this here, it'll end up spamming the sentry logs, maybe make these qDebug so it's not showing in release builds.

Reply:
It's copy-paste from the layer observer. Changed to qDebug anyways here. We should not change layer observer for now I would say.